tensorflow/python/keras/layers/convolutional.py

```diff
@@ -151,7 +167,14 @@ def __init__(self, rank,
   def build(self, input_shape):
     input_shape = tensor_shape.TensorShape(input_shape)
     input_channel = self._get_input_channel(input_shape)
-    kernel_shape = self.kernel_size + (input_channel, self.filters)
+    if input_channel % self.groups != 0:
+      raise ValueError(
+          'The number of input channels must be evenly divisible by the number '
+          'of groups. Received groups={}, but the input has {} channels '
+          '(full input shape is {}).'.format(self.groups, input_channel,
+                                             input_shape))
+    kernel_shape = self.kernel_size + (input_channel // self.groups,
+                                       self.filters)
 
     self.kernel = self.add_weight(
```

Comment:
I'm a little confused here, dont we need `self.groups` # of kernels each of shape `(kernel_size,) + (input_channel // self.groups, self.filters)`?

Reply:
As far as I know most frameworks follow the convention that the shape of the weights is `(*kernel_size, input_channel // groups, output_channels)` so that `filters==output_channels` regardless of the number of groups used. E.g. PyTorch does the same [here](https://github.com/pytorch/pytorch/blob/82d58ed484eb04f894475de0551055f7e070f481/torch/nn/modules/conv.py#L47-L48).
I personally think of `filters` as the number of output channels of a convolution, so adopting the convention that `filters` are the number of channels per group doesn't seem natural to me, but please correct me if I am wrong, this seems to be more a question of definition.
For me figure 3 in https://arxiv.org/pdf/1611.05431.pdf suggests to prefer `filters==output_channels` as well.

Reply:
Right, so `filters==output_channels`.
What I don't understand here is, let's say we want to map a (55, 55, 32) conv input to (27, 27, 64) conv output, with `groups=2`. So apparently we need 2 different kernels, both are with shape (3, 3, 16, 64). The first kernel will map the first half (55, 55, 16) to (27, 27, 64), and the second kernel will map the second half (55, 55, 16) to (27, 27, 64). And the final result would be the sum of the two.
So either we need 2 kernels in 1 conv layer, or we need 2 conv layers. Am I missing something here? (Feel free to point out where I'm wrong)

Reply:
`filters` gets split into `groups` by the convolutional algorithm. If you would have multiple convolutions mapping to 64d, what would be the end result? 128d instead of 64d for `groups==2`? As far as I know this is handled by the convolution `tf.nn.conv*` or cudnn.

Reply:
@tanzhenyu The grouped convolution computes splits  in the output channel dimension which are then concatenated to combine the result. Your example above would change to:

The first kernel will map the first half (55, 55, 16) to (27, 27, ~~64~~**32**), and the second kernel will map the second half (55, 55, 16) to (27, 27, ~~64~~**32**). And the final result would be the ~~sum~~**concatenation** of the two.

The implementation of `tf.nn.conv2d` handles this computation either using XLA or CUDNN. This is the reason why adding grouped convolution support to Keras only requires a few additional lines that change the kernel shape. To be fair the fact that `tf.nn.conv2d` computes a grouped convolution depending on the shapes of the inputs and weights is a bit confusing in my opinion.

Consider the following low level TensorFlow implementation of your example that showcases the computation in relation to a simple loop based implementation and the relation of a grouped convolution to a depthwise:

```python
import numpy as np
import tensorflow as tf


@tf.function(experimental_compile=True)  # Needed for grouped conv CPU support
def conv_2d(inputs, weights, strides=[1, 1, 1, 1], padding="VALID"):
    return tf.nn.conv2d(inputs, weights, strides=strides, padding=padding)


# loop based convolution

groups = 2
inputs = tf.random.normal(shape=[1, 55, 55, 32], dtype=tf.float64)
weights = tf.random.normal(shape=[3, 3, 32 // groups, 64], dtype=tf.float64)

outputs = conv_2d(inputs, weights, strides=[1, 2, 2, 1])

grouped_inputs = tf.split(inputs, groups, axis=-1)
grouped_weights = tf.split(weights, groups, axis=-1)
outputs_ref = tf.concat(
    [
        conv_2d(inputs, weights, strides=[1, 2, 2, 1])
        for inputs, weights in zip(grouped_inputs, grouped_weights)
    ],
    axis=-1,
)

np.testing.assert_allclose(outputs.numpy(), outputs_ref.numpy())

# depthwise

inputs_dw = tf.random.normal(shape=[1, 55, 55, 64], dtype=tf.float64)
weights_dw = tf.random.normal(shape=[3, 3, 64, 1], dtype=tf.float64)
weights_grouped_dw = tf.reshape(weights_dw, [3, 3, 1, 64])

outputs_dw = conv_2d(inputs_dw, weights_grouped_dw, strides=[1, 2, 2, 1])
outputs_dw_ref = tf.nn.depthwise_conv2d(
    inputs_dw, weights_dw, strides=[1, 2, 2, 1], padding="VALID"
)

np.testing.assert_allclose(outputs_dw.numpy(), outputs_dw_ref.numpy())
```

Reply:
Ok I'm assuming this will end up calling tf.nn.conv2d with `inputs=(H,W, IN_CHANNEL)` and `filters=(IN_CHANNEL/GROUP, K, K, OUT_CHANNEL)` and under the hood (at kernel level) it will infer the group and get multiple OUT_CHANNEL/GROUP conv outputs and concatenate before returning.

Reply:
Oh I refreshed the page before your comment. Yeah this seems legit.

Reply:
@lgeiger Though to be more concise, I think when we call `tf.nn.conv2d`, the inputs don't need to be splitted, because we need to infer groups from input channel and kernel size?
(The docstring of that op needs to be updated as well)

Reply:
> Though to be more concise, I think when we call `tf.nn.conv2d`, the inputs don't need to be splitted, because we need to infer groups from input channel and kernel size?

That's correct and that's what this PR should do as well, since we don't actually change `Layer.call`.